dvc/stage.py

```diff
@@ -671,9 +669,18 @@ def dump(self):
         logger.debug(
             "Saving information to '{file}'.".format(file=relpath(fname))
         )
-        d = self.dumpd()
-        apply_diff(d, self._state)
-        dump_stage_file(fname, self._state)
+        state = self.dumpd()
+
+        # If we have stage file we apply diffs instead of overwriting it
+        # to preserve comments, string formatting and meta key.
+        if self._stage_text is not None:
+            saved_state = parse_stage_for_update(self._stage_text, fname)
+            if "meta" in saved_state:
+                state["meta"] = saved_state["meta"]
```

Comment:
why is some special handling required for meta here? (vs just a single `apply_diff` before)

Reply:
Because `state`, which we get from `self.dumpd()` does not have `meta` key. So `apply_diff()` will remove it in the target structure. The alternative would be to store `meta` in some key and then include it into `.dumpd()`, which is more hustle.

Reply:
But why didn't we have the same (seemingly very ad-hoc and fragile logic) before? We were using dump before right? and were applying diff on top of it?

Reply:
I see. Before it was part of the `dumpd`. I missed that a line was removed in `dumpd` as well.

Still looks fragile that we have one more place that manipulates with all these attributes. It feels that it would be better to follow the "regular" way in this file and make the `meta` a regular stage attribute. So, that we don't have a separate path to handle just it.

Also, wondering if `is_cached` can be optimized considering that we are saving the state.

Reply:
The regular way now will mean we will need to store `meta` on some attribute, then return it in `.dumpd()` then let it be applied to the same meta, but with comments etc inside `apply_diff()`.

I am not sure that is a good idea, I don't want to add an another attribute, which is never used in fact. And even less I want to add another constructor param.